### DISCUSSION TITLE
Can't next build from performance-relayer.js location is not defined

### DISCUSSION THREAD

**Discussion by gillerst1** [vercel/next.js#20383] [Help]
ReferenceError: location is not defined
    at Object.<anonymous> (/PATH/node_modules/next/dist/client/performance-relayer.js:1:135)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/PATH/node_modules/next/dist/client/index.js:1:1406)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10) {
  type: 'ReferenceError'

What do I do??

**Comment by logandonley** [comment]
I'm encountering this issue as well. I've followed the references in my stack trace and added some if (process.browser) {} blocks in the appropriate files in the node_modules directory and I was able to build.

I haven't setup any of the performance metrics functionality but it seems like that component still gets touched during build since it's failing due to using windows & location calls that aren't being checked that they are in the browser.

**Reply by jcrangon** [reply]
I have the same error and I really do not understand where it comes from.
Could you explain exactly where youput the process.browser block?

**Reply by jcrangon** [reply]
I found a solution by creating a fresh copy of the project, then removing all created files except _app, _document and index.js, then adding them one by one and building each time until I found the incriminated file. In my case it was an export function that was working and somehow, it started to generate this error.. but it is still not clear why.

**Reply by keepeek-rd** [reply]
I had the same problem and it came from this piece of code in _page:
const router = useRouter ()
   useEffect (() => {
     const handleRouteChange = (url) => {
       gtag.pageview (url)
     }
     router.events.on ('routeChangeComplete', handleRouteChange)
     return () => {
       router.events.off ('routeChangeComplete', handleRouteChange)
     }
   }, [router.events])

which I replaced by:
Router.events.on ("routeChangeComplete", (url) => gtag.pageview (url));

**Comment by greghesp** [comment]
I had the same build error, it turns out that when I had used `router.push()`, my IDE had automatically imported router from `next/client` rather than me adding in the `useRouter` hook

**Reply by njoguamos** [reply]
I had the same problem. Your solution worked.

**Reply by savisafe** [reply]
working, my problem code router.back()

**Reply by Kazimir42** [reply]
thx :)

**Reply by ramyozi** [reply]
same problem, i had to ctrl shift f and look for all instances of import with next/client

**Reply by hoa-nguyen-bk** [reply]
but it's still not working with me

**Reply by RemoteRacoon** [reply]
thank you so muck for idea!

**Comment by robert-yarborough** [comment]
I'm receiving the same error message. But in my context, it's after utilizing 'dynamic-routes'. Which as you all remember, nextjs exposes the specified id objects as routes, in getStaticPaths. 
I'm going to research and eventually find the proper solution, but this really blows. 
There's no real reference in the nextjs docs or forums.

**Reply by robert-yarborough** [reply]
SOLVED: In my case, my text-editor, webstorm. Automatically imported a render module from nextjs.
 import {render} from "next/client";
Immediately after deleting it, my program worked smoothly. 

So as @greghesp  pointed out earlier in this thread, check your IDE imports. And make sure you don't have any unwanted imports. 
CHEERS 🥂 🍻 🍺

**Reply by IanDex** [reply]
I love you, in my case I import router

import {router} from "next/client"

**Comment by ismoil793** [comment]
There is a solution at Stackoverflow please check out the link below:
https://stackoverflow.com/questions/65819812/location-is-not-defined-error-in-react-next-js/67670615#67670615

**Comment by chelizichen** [comment]
don't use ` import {router} from 'next/client', it will make a problem in windows and it test success in macos
